---
title: "Kidney"
output: html_document
date: "2023-07-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Installations and imports

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(hdf5r)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggplot2)
library(cowplot)
```

## Data

Reading the data
TODO: both for raw and filtered, what are the steps to get filtered?

```{r}
# load the RNA and ATAC data
setwd("~/R_scripts")
counts <- Read10X_h5("../datasets/kidney_10x/H_Kidney_Cancer_Chromium_Nuc_Iso_vs_SaltyEZ_vs_ComplexTissueDP_filtered_feature_bc_matrix.h5")
fragpath <- "../datasets/kidney_10x/fragments.tsv.gz"
```

Creating the Seurat object and annotating the data. 
Selecting scATAC subset

```{r}
# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevels(annotation) <- paste0('chr', seqlevels(annotation))

# create a Seurat object containing the RNA adata
kidney <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)

# create ATAC assay and add it to the object
kidney[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotation
)

kidney
```
```{r}
saveRDS(kidney, file='./temp_data/kidney_f_nf_annotated.RDS')
```

## Quality control

Checking the distribution of the data:

```{r}
DefaultAssay(kidney) <- "ATAC"

# compute nucleosome signal score per cell
kidney <- NucleosomeSignal(kidney)
# compute TSS enrichment score per cell
kidney <- TSSEnrichment(kidney, fast=FALSE)

# add blacklist ratio and fraction of reads in peaks
kidney$pct_reads_in_peaks <- kidney$peak_region_fragments / kidney$passed_filters * 100
kidney$blacklist_ratio <- kidney$blacklist_region_fragments / kidney$peak_region_fragments
```

```{r}
saveRDS(kidney, file='./temp_data/kidney_f_nf_annotated_withQC.RDS')
```

Some plots:

Useful plot to understand the relationship between TSS and peak counts (or any other variables)

```{r}
DensityScatter(kidney, x = 'nCount_peaks', y = 'TSS.enrichment', log_x = TRUE, quantiles = TRUE)
```

Inspecting the TSS enrichment scores by grouping the cells based on the score and plotting the accessibility signal over all TSS sites

```{r}
kidney$high.tss <- ifelse(kidney$TSS.enrichment > 3, 'High', 'Low')
TSSPlot(kidney, group.by = 'high.tss') + NoLegend()
```

Fragment length periodicity for all the cells, and group by cells with high or low nucleosomal signal strength

```{r}
kidney$nucleosome_group <- ifelse(kidney$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
FragmentHistogram(object = kidney, group.by = 'nucleosome_group')
```

Distributions of all metrics:

```{r}
VlnPlot(
  object = kidney,
  features = c('nCount_peaks', 'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal', 'pct_reads_in_peaks'),
  pt.size = 0.1,
  ncol = 5
)
```

### Filtering 
Filtering out low quality cells and genes. Here we are selecting constants based on the distributions above:

```{r}
# filter out low quality cells
kidney <- subset(
  x = kidney,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 25000 &
    nCount_ATAC > 1000 &
    nCount_RNA > 1000 &
    nucleosome_signal < 2 &
    TSS.enrichment > 1
)
kidney
```

Checking the distributions of all metrics again:

```{r}
VlnPlot(
  object = kidney,
  features = c('nCount_peaks', 'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal', 'pct_reads_in_peaks'),
  pt.size = 0.1,
  ncol = 5
)
```

## Normalisation 

CHECK HOW THEY ALL WORK!!!

Signac performs term frequency-inverse document frequency (TF-IDF) normalization. This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks. FORMULA??

```{r}
kidney <- RunTFIDF(kidney)
```

Then we do feature selection - selecting top n% features, we can set then by q = percent to remove.
Features used for dimensional reduction are automatically set as VariableFeatures() for the Seurat object by this function.
Here we will use all????

```{r}
kidney <- FindTopFeatures(kidney, min.cutoff = 'q0')
```

Now we do dimension reduction - here we run singular value decomposition (SVD) on the TD-IDF matrix, using the features (peaks) selected above. This returns a reduced dimension representation of the object.
WHATS THE METHOD AND FORMULA??

```{r}
kidney <- RunSVD(kidney)
```

Correlation between components:

```{r}
DepthCor(kidney)
```



## Peak calling 

```{r}
# call peaks using MACS2
peaks <- CallPeaks(kidney)

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(kidney),
  features = peaks,
  cells = colnames(kidney)
)

# create a new assay using the MACS2 peak set and add it to the Seurat object
pbmc[["peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragpath,
  annotation = annotation
)
```


```{r}

```


```{r}

```


```{r}

```


```{r}

```

```{r}

```

```{r}

```

```{r}

```

